src/ros_filter.cpp

```diff
@@ -381,23 +384,37 @@ namespace RobotLocalization
   template<typename T>
   void RosFilter<T>::integrateMeasurements(const double currentTime)
   {
+    boost::mutex::scoped_lock lock(measurementQueueMutex_);
+
     RF_DEBUG("------ RosFilter::integrateMeasurements ------\n\n"
              "Integration time is " << std::setprecision(20) << currentTime << "\n"
              << measurementQueue_.size() << " measurements in queue.\n");
 
     // If we have any measurements in the queue, process them
-    if (!measurementQueue_.empty())
+    if(measurementQueueEmpty_.timed_wait(lock,boost::posix_time::milliseconds(filter_.getSensorTimeout()*1000)))
     {
-      while (!measurementQueue_.empty())
+      if(!measurementQueue_.empty())
       {
-        Measurement measurement = measurementQueue_.top();
-        measurementQueue_.pop();
 
-        // This will call predict and, if necessary, correct
-        filter_.processMeasurement(measurement);
+        double lastMessageTime;
+        while (!measurementQueue_.empty())
+        {
+          Measurement measurement = measurementQueue_.top();
+          measurementQueue_.pop();
+          lock.unlock();
+          // This will call predict and, if necessary, correct
+          filter_.processMeasurement(measurement);
+          lastMessageTime = measurement.time_;
+          lock.lock();
+          if (!measurementQueue_.empty())
```

Comment:
Would it make sense to grab the data necessary for this logic after line 403? In other words, between lines 403 and 404, insert this:

queueEmpty = measurementQueue_.empty()
if(!queueEmpty)
{
  nextMeasurementTime = measurementQueue_.top().time_;
}

Then, you can avoid using the lock on lines 408 and 412, and use the queueEmpty variable on line 409, and use the nextMeasurementTime variable on line 410. It'll save you a lock and unlock.